firebase-firestore/src/main/java/com/google/firebase/firestore/encoding/MapEncoder.java

```diff
@@ -27,10 +28,15 @@ public interface MapEncoder {
   // Encodes an custom object to a nested map of firestore primitive data types
   Map<String, Object> encode(Object value);
 
+  // Decodes a map to custom object
+  <T> T decode(Map<String, Object> data, Class<T> valueType, DocumentReference docRef);
+
   // Uses reflection to determine if this object is suitable to be encoded with the Kotlin
   // serialization plugin
   boolean isAbleToBeEncoded(Object value);
 
+  <T> boolean isAbleToBeDecoded(Class<T> valueType);
```

Comment:
Do we need 2 methods? is it possible that some future encoder would want to support only encoding or only decoding? also do we want it to be possible?

How about having just one method that would indicate support for both directions? e.g.

```java
boolean supports(Class<?> cls);
```

Reply:
I brought up my concern to Rodrigo that there are some cases, where one group of people saving to firestore, and another group of people get data from firestore. It will be cool/useful if we give customer the option to choose which encoder/decoder they want to use at runtime.  [I do heard that the company Strip is saving their enum field to firestore in BigCamel format, and try to get them back as smallCamel ]

Rodrigo is agree with you that we should not give customer too much freedom, and it makes customer life easier as they have our Kotlin extension lib, they can use the Kotlin serialization style in both encoding and decoding. So eventually, we are following your `supports()` suggestion. Thank you.